### PR TITLE
[improve][misc] Add LongArrayAckSets for allocation-free ack set operations

### PR DESCRIPTION
**Pull request by Shawyeok**
### Motivation

Several batch-ack paths allocate a `BitSet`/`BitSetRecyclable` just to compute a cardinality or an
AND over raw `long[]` word arrays. These allocations are unnecessary — the results can be computed
directly with `Long.bitCount` and bitwise ops, and some call sites were doing more scanning than
they needed to.

### Modifications

**New `LongArrayAckSets` utility class** (`pulsar-common`), operating on the "long array" bit-set
representation produced by `BitSet#toLongArray()` / consumed by `BitSet#valueOf(long[])`:

| Method | Description |
|--------|-------------|
| `cardinality(long[])` | Popcount of a word array — no `BitSet` alloc |
| `isEmpty(long[])` | Returns `false` at the first non-zero word, instead of popcounting the whole array |
| `intersect(long[], long[])` | Element-wise AND, trailing zero words trimmed to match `BitSet#toLongArray()`'s canonical form |
| `cardinalityOfIntersection(long[], long[])` | Popcount of AND without allocating a result array |
| `cardinalityOfDifference(long[], long[])` | Popcount of `ackSet1 & ~ackSet2` in one pass, instead of `cardinality(ackSet1) - cardinalityOfIntersection(ackSet1, ackSet2)` (two passes) |

**Refactored call sites**:

- `Consumer` (broker) — `computeAckedCount` uses `cardinalityOfDifference` (single pass instead of two); other cardinality checks use `LongArrayAckSets.cardinality`
- `EntryBatchIndexesAcks` (broker) — cardinality check uses `LongArrayAckSets.cardinality`
- `PositionAckSetUtil` (managed-ledger) — `andAckSet` delegates to `LongArrayAckSets.intersect`; `isAckSetEmpty` uses the short-circuiting `LongArrayAckSets.isEmpty` instead of a full cardinality scan

**Tests**: `LongArrayAckSetsTest` covers all five methods (empty arrays, all-zero, all-set, mixed,
asymmetric lengths, no/full/partial overlap).

### Verifying this change

- [x] Make sure that the change passes the CI checks.

```bash
./gradlew :pulsar-common:test --tests org.apache.pulsar.common.util.LongArrayAckSetsTest
./gradlew :managed-ledger:test --tests org.apache.bookkeeper.mledger.util.PositionAckSetUtilTest
./gradlew :pulsar-broker:compileJava
```

### Does this pull request potentially affect one of the following parts:

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### CODE REVIEW

**Review by lhotari — CHANGES_REQUESTED** [review]
Please check these findings of a local Claude Code review:

The cardinality-based refactors in this PR are exactly equivalent to the code they replace. However, two of the rewrites are **not** strictly equivalent in one respect: **handling of trailing all-zero words**. The old code trimmed them (via `BitSet.valueOf` / `toLongArray`); the new code does not.

### 1. `intersect` no longer trims trailing zero words — breaks the "canonical ack-set" invariant the old `andAckSet` upheld

`AckSetUtil.java` (`intersect`) and `PositionAckSetUtil.java` (`andAckSet`)

The old `andAckSet` ended with `thisAckSet.toLongArray()`, which returns `Arrays.copyOf(words, wordsInUse)` — trailing all-zero words are dropped. The new `intersect` returns a `min(len1, len2)`-length array **without trimming**. When two multi-word ack sets AND to a zero *high* word (very possible for batches > 64 indexes, e.g. `[w0, 0b01] & [w0, 0b10]` → `[.., 0]`), the two implementations return arrays of different length/shape:

- old: `[w0]` (trimmed)
- new: `[w0, 0]` (not trimmed)

This result is stored back into transaction pending-ack state via `andAckSet(pos1, pos2)` → `setAckSet(...)` (`PendingAckHandleImpl`) and into `AbstractBaseDispatcher`. Most downstream consumers are trailing-zero-safe (`isAckSetEmpty`/`cardinality*` popcount to 0; `BitSet.valueOf` re-trims on read/serialize), so persistence is unaffected — **but `isAckSetOverlap` is not safe** (see finding 2), so a non-canonical array can flow into it and change the result.

**Suggestion:** trim trailing zero words in `intersect` (to match `toLongArray()`), or explicitly document/enforce that callers only pass and store canonical (trimmed) arrays.

### 2. `isAckSetOverlap` rewrite is trailing-zero-sensitive where the old code was not

`PositionAckSetUtil.java` (`isAckSetOverlap`)

The old code did `BitSetRecyclable.valueOf(x)` (which trims trailing zero words) before `flip`/`and`, so a trailing all-zero word was ignored. The new loop `~(a[i] | b[i]) != 0` over `min(len)` treats a shared trailing zero word as "acked in both" → returns `true` where the old code returned `false`. Concretely:

```java
isAckSetOverlap(new long[]{-1L, 0L}, new long[]{-1L, 0L})
// old = false, new = true
```

Within-word high padding is handled identically by old and new (both operate on full 64-bit words), so the *only* divergence is whole trailing zero words — which is exactly what finding 1 can now feed in. The two changes interact: an `intersect` result like `[w0, 0]` compared against another length-≥2 set can now flip an overlap decision (this drives transaction double-ack / conflict detection).

Note the existing `PositionAckSetUtilTest` compares via `BitSetRecyclable.valueOf(...)` (which trims), so it **cannot catch** this — the divergence is untested. Please add targeted tests: a multi-word AND whose high word is zero, and an `isAckSetOverlap` case with a trailing zero word.

### 3. `AckSetUtil` javadoc / style nits

`AckSetUtil.java`

- `cardinalityOfIntersection`: the `@return` text references `firstWords & secondWords`, but the params are named `set1`/`set2`; and the `@param set2` line has a stray extra space / misaligned indentation.
- Methods are declared `public static` inside a Lombok `@UtilityClass`, which already forces `static` — the explicit modifier is redundant (harmless, but inconsistent with the annotation's intent).

**Comment by Shawyeok** [comment]
@lhotari Thanks for elaborate review, here is what I addressed:

> 1. intersect no longer trims trailing zero words — breaks the "canonical ack-set" invariant the old andAckSet upheld

Fixed, now `AckSetUtil#intersect` is aligned with `BitSet#toLongArray` for trimming the trailing zero word automatically.

> 2. `isAckSetOverlap` rewrite is trailing-zero-sensitive where the old code was not

I've restored the change of the method `PositionAckSetUtil#isAckSetOverlap`, because without the batch size, a trailing zero word is ambiguous — `BitSetRecyclable.valueOf` trimming means `[-1L, 0L]` ("bits 64–127 acked") is treated as if those bits don't exist. Maybe we can fix this later in a dedicated PR.

> 3. AckSetUtil javadoc / style nits

javadoc fixed, I've removed the Lombok annotation `@UtilityClass`

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.util;
+
+/**
+ * Utility methods for operating on ack-set word arrays without allocating a {@code BitSet} instance.
+ */
+public class AckSetUtil {
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java]: This class could be called `LongArrayBitSets`. In some code locations, the concept for a BitSet in `long[]` word representation is called "long array".  I guess this concept comes from `BitSet.toLongArray` method.

I prefer a plural format for the class name of a class that contains static methods for handling specific concepts. That's why I'd avoid the `Util` suffix. It also directs later maintainers so that this class doesn't become a container for all sorts of utility methods.

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.util;
+
+/**
+ * Utility methods for operating on ack-set word arrays without allocating a {@code BitSet} instance.
+ */
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java:26]: This could be generalized so that it's about operating on "a long array containing a little-endian representation of all the bits in a bit set". There could be javadoc links to `java.util.BitSet#toLongArray` and `java.util.BitSet#valueOf(long[])` as a reference.

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * <p>Extra words in the longer array are implicitly ANDed with zero and therefore contribute no set bits.
+     * Trailing all-zero words are trimmed from the result, so it is in the same canonical form produced by
+     * {@link java.util.BitSet#toLongArray()}.
+     *
+     * @param set1 a long array containing a little-endian representation of a sequence of bits
+     * @param set2 a long array containing a little-endian representation of a sequence of bits
+     * @return a new array representing the intersection of the two bit sets, with trailing zero words trimmed
+     */
+    public static long[] intersect(long[] set1, long[] set2) {
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java]: perhaps it would be more accurate to rename the parameters to `bitSet1` and `bitSet2`. the `long[]` type already makes it clear that these are "long array bit sets"

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ */
+public class AckSetUtil {
+
+    /**
+     * Returns the number of bits set to {@code true} in the given words.
+     *
+     * @param words a long array containing a little-endian representation of a sequence of bits
+     * @return the number of bits set to {@code true}
+     */
+    public static int cardinality(long[] words) {
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java]: words -> bitSet, to be consistent with the "long array bit set" concept

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+     * Returns the number of bits set to {@code true} after applying a logical <b>AND</b> to the given words.
+     *
+     * <p>When the arrays differ in length, extra words in the longer array are treated as zero (i.e. the AND
+     * result for those positions is zero and contributes nothing to the cardinality).
+     *
+     * @param set1 a long array containing a little-endian representation of a sequence of bits
+     * @param set2 a long array containing a little-endian representation of a sequence of bits
+     * @return the number of bits set to {@code true} in {@code set1 & set2}
+     */
+    public static int cardinalityOfIntersection(long[] set1, long[] set2) {
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/AckSetUtil.java]: rename parameters

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+public class AckSetUtil {
+
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java:31]: Add a private constructor so that this class cannot be instantiated

**Review by lhotari — COMMENTED** [review]
A comment about renaming to `AckSetUtil` to `LongArrayBitSets`.

**Comment by lhotari** [comment]
After addressing comments, please also update the PR title and description. title could be something like "Add LongArrayBitSets for allocation free bit set operations and use it for ack set handling" etc.

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java (hunk 6 above)
Comment (Shawyeok) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java:31]: > This class could be called LongArrayBitSets. In some code locations, the concept for a BitSet in long[] word representation is called "long array". I guess this concept comes from BitSet.toLongArray method.

Thanks for review and suggestion, I'd rather use `LongArrayAckSets` here, because AckSet is a very special usage of BitSet in pulsar, WDYT?

Comment (Shawyeok) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java:31]: All addressed, feel free to take a look.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/PositionAckSetUtil.java [hunk 7]
```diff
@@ -54,20 +55,11 @@ public static void andAckSet(Position currentPosition, Position otherPosition) {
...
-        return ackSet;
+        return LongArrayAckSets.intersect(firstAckSet, secondAckSet);
     }
 
     public static boolean isAckSetEmpty(long[] ackSet) {
-        BitSetRecyclable bitSet =  BitSetRecyclable.create().resetWords(ackSet);
-        boolean isEmpty = bitSet.isEmpty();
-        bitSet.recycle();
-        return isEmpty;
+        return LongArrayAckSets.cardinality(ackSet) == 0;
```
Comment (Denovo1998) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/PositionAckSetUtil.java:62]: Since this call only needs an emptiness check, computing the full cardinality scans and popcounts every word even after a non-zero word has been found. This is on the dispatcher path and the purpose of the PR is to reduce ack-set overhead. Could LongArrayAckSets expose an isEmpty(long[]) helper that returns false at the first non-zero word, and use that here?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java [hunk 8]
```diff
@@ -804,23 +803,15 @@ private long computeAckedCount(MessageIdData msgId, Position position, Consumer
...
         }
-        BitSetRecyclable cursorBitSet = BitSetRecyclable.create().resetWords(cursorAckSet);
-        int lastCardinality = cursorBitSet.cardinality();
-        BitSetRecyclable givenBitSet = BitSetRecyclable.create().resetWords(ackSets);
-        cursorBitSet.and(givenBitSet);
-        givenBitSet.recycle();
-        int currentCardinality = cursorBitSet.cardinality();
-        cursorBitSet.recycle();
+        int lastCardinality = LongArrayAckSets.cardinality(cursorAckSet);
+        int currentCardinality = LongArrayAckSets.cardinalityOfIntersection(cursorAckSet, ackSets);
```
Comment (Denovo1998) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:809]: This result can be computed with one scan instead of scanning cursorAckSet twice. For every bit, C - (C & A) is equivalent to C & ~A, so:

```
cardinality(cursor) - cardinality(cursor & ack)
= cardinality(cursor & ~ack)
```

Here, cursor & ~ack directly represents indexes that were unacked in the cursor and become acked by this acknowledgment. For unequal array lengths, words beyond ackSets.length must still contribute their full cursor cardinality, because the missing ack words are treated as zero by the previous BitSet.and implementation. 

Could this be implemented as a single-pass helper such as cardinalityOfDifference(cursorAckSet, ackSets)?

```java
return LongArrayAckSets.cardinalityOfDifference(cursorAckSet, ackSets);

public static int cardinalityOfDifference(long[] ackSet1, long[] ackSet2) {
    int sum = 0;
    int commonLength = Math.min(ackSet1.length, ackSet2.length);

    for (int i = 0; i < commonLength; i++) {
        sum += Long.bitCount(ackSet1[i] & ~ackSet2[i]);
    }

    for (int i = commonLength; i < ackSet1.length; i++) {
        sum += Long.bitCount(ackSet1[i]);
    }

    return sum;
}
```

**Review by Denovo1998 — DISMISSED** [review]
@lhotari @Shawyeok Need a JMH benchmark or profiling evidence?

**Review by lhotari — COMMENTED** [review]
Re-reviewed at `4ad88cd`. The only commit since my last pass is the rename, and it is exactly what was asked for: `AckSetUtil` → `LongArrayAckSets`, parameters renamed to `ackSet` / `ackSet1` / `ackSet2`, javadoc that now names the representation explicitly, and a private constructor so the utility class cannot be instantiated. The algorithms are byte-identical apart from renamed identifiers — no logic changed.

I also re-checked the two findings from my earlier review rather than relying on the threads being marked resolved, and both are genuinely fixed at this head:

- **`intersect` trailing-zero trimming** — now trimmed, via the `while (len > 0 && (ackSet1[len - 1] & ackSet2[len - 1]) == 0) len--;` loop, and the javadoc states the result is in the canonical form produced by `BitSet.toLongArray()`.
- **`isAckSetOverlap` trailing-zero sensitivity** — resolved by leaving the method on `BitSetRecyclable.valueOf(...)` rather than keeping the loop rewrite. `BitSetRecyclable.valueOf` trims, so the `[-1L, 0L]` vs `[-1L, 0L]` divergence I gave as the concrete case no longer arises.

That clears everything on my side.

**On the CI failure — it is not this PR.** The only failing check is `Flaky tests suite`, which failed after 19 seconds with `The action gradle/actions/setup-gradle@39e147cb… is not allowed in apache/pulsar because all actions must be from a repository owned by your enterprise…`. That is the repository's allowed-actions policy, not the code, and the run dates from 2026-07-22. Nothing to fix here; a rerun on current `master` should sort it out.

**@Denovo1998 — on your question about benchmark evidence.** I think a small JMH benchmark is worth it and cheap here, but full profiling is not needed. These are pure functions over `long[]` with no setup, so a microbenchmark is a few lines and gives a defensible number. The allocation claim in the title is structurally evident from the diff — the `BitSetRecyclable.valueOf(...)`/`toLongArray()` round-trip is gone from these paths — but "allocation-free" and "faster" are different claims, and only the second one needs measuring.

**Your two open suggestions both look right to me, and I would like to see them in this PR.**

1. `isEmpty(long[])` short-circuiting at the first non-zero word, used by `PositionAckSetUtil.isAckSetEmpty` — agreed. That method currently computes `LongArrayAckSets.cardinality(ackSet) == 0`, which popcounts every word to answer a question the first set bit already settles. It is on the dispatcher path and the PR is specifically about ack-set overhead, so this is squarely in scope.
2. `cardinalityOfDifference(cursorAckSet, ackSets)` — I checked the identity and it holds. `C` partitions disjointly into `C & A` and `C & ~A`, so `cardinality(C) - cardinality(C & A) == cardinality(C & ~A)`, and your handling of the unequal-length tail is right: words past `ackSets.length` are treated as zero, `C & ~0 == C`, so they contribute their full cursor cardinality — which is what the second loop in your sketch does. It replaces two scans with one on the same path.

Neither is a correctness problem with what is here now, so this is not a blocker from me — but they are the kind of thing this PR exists to do, and the second one comes with a working implementation already written.

**Review by lhotari — COMMENTED** [review]
Re-checked at `4accdf7d`. Nothing has moved on the author's side since my last pass — the only new commit is the `master` merge from the **Update branch** button, and it is clean. The merge changes `Consumer.java` only to pick up the packed-long `PendingAcksMap` rework from #26030 (`removePendingAckAndGet` → `removePendingAckAndGetRemainingUnacked`), which does not touch any of the four `LongArrayAckSets` call sites this PR introduces. `LongArrayAckSets`, `LongArrayAckSetsTest`, `PositionAckSetUtil`, `PositionAckSetUtilTest`, `EntryBatchIndexesAcks` and `BitSetRecyclable` are byte-identical to `4ad88cd`, so everything I verified last round still holds at this head.

CI is green on the merged head, which settles the earlier failure — it was the repository's allowed-actions policy, not this change.

**The one thing still outstanding from July is the PR description.** On 3 July I asked for the title and the description to be updated after the rename. The title was updated on 6 July; the body was not, and it still describes a state the branch left behind. That matters more than it usually would, because the PR body becomes the squashed commit message and would land in `master` as written. Concretely:

- It introduces "a new `AckSetUtil` utility class" and a "dedicated `AckSetUtilTest`". Both are `LongArrayAckSets` / `LongArrayAckSetsTest` now, and the `./gradlew … --tests org.apache.pulsar.common.util.AckSetUtilTest` line under *Verifying this change* names a class that does not exist on the branch.
- It says `intersect` returns a result of "length = `min(len1, len2)`". After the trimming fix that is no longer true: trailing all-zero words are dropped so the result is in the canonical `BitSet.toLongArray()` form — which is precisely the invariant the change now rests on. See [`LongArrayAckSets.java:57-67`](https://github.com/apache/pulsar/blob/4accdf7dfb1e6d80ca4ff90cd171e7adeddb1f7a/pulsar-common/src/main/java/org/apache/pulsar/common/util/LongArrayAckSets.java#L57-L67).
- It says `isAckSetOverlap` was "rewritten with bitwise ops … eliminates two `BitSetRecyclable` allocs plus two full `flip` passes". That rewrite was reverted in `3667187e`, and the method is unchanged from `master`:

https://github.com/apache/pulsar/blob/4accdf7dfb1e6d80ca4ff90cd171e7adeddb1f7a/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/PositionAckSetUtil.java#L30-L44

The rest of *Modifications* is accurate — I re-checked the four `Consumer` sites and the one in `EntryBatchIndexesAcks` against `master` at this head.

**Still open from @Denovo1998's review**, and unchanged since I said on 28 August that I agreed with both: the `isEmpty(long[])` short-circuit behind `isAckSetEmpty` ([`PositionAckSetUtil.java:61-63`](https://github.com/apache/pulsar/blob/4accdf7dfb1e6d80ca4ff90cd171e7adeddb1f7a/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/PositionAckSetUtil.java#L61-L63)), and the single-pass `cardinalityOfDifference` that would replace the two scans in `computeAckedCount` ([`pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:808-810`](https://github.com/apache/pulsar/blob/4accdf7dfb1e6d80ca4ff90cd171e7adeddb1f7a/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java#L808-L810)). Neither is a correctness problem with what is here, so taking them or pushing back on them both work — but the PR still carries a changes requested decision, and a reply either way is what would move it.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/PositionAckSetUtil.java (hunk 7 above)
Comment (Shawyeok) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/PositionAckSetUtil.java]: Sorry for the late reply. Good catch, fixed. Added `LongArrayAckSets.isEmpty(long[])`, which returns `false` as soon as it hits a non-zero word, and switched `isAckSetEmpty` to use it instead of `cardinality(ackSet) == 0`.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java (hunk 8 above)
Comment (Shawyeok) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java]: Sorry for the late reply. Added `LongArrayAckSets.cardinalityOfDifference(long[], long[])` exactly as you described — single pass over `ackSet1[i] & ~ackSet2[i]`, with words beyond the shorter array's length treated as zero — and switched `computeAckedCount` to use it instead of the two-scan `cardinality(cursor) - cardinalityOfIntersection(cursor, ack)`.

**Comment by Shawyeok** [comment]
@lhotari @Denovo1998 
Sorry for the late response, I've been busy since August. I've added `isEmpty` and `cardinalityOfDifference` methods and corresponding tests, also updated the pr description to reflect the latest change.

**Review by lhotari — APPROVED** [review]
LGTM. The new `isEmpty` and `cardinalityOfDifference` helpers address both remaining review points, including the unequal-length case where unmatched cursor words must contribute their full bit count. The updated description now matches the implementation and the test names.
